### PR TITLE
Remove patch that uses waagent for fabric negotiation

### PR DESCRIPTION
**Pull request by trstringer**
Moving the [merge request](https://code.launchpad.net/~trstringer/cloud-init/+git/cloud-init/+merge/375217) from Launchpad to GitHub.

Description from Launchpad:

> This merge request removes the patch from the Xenial branch that alters cloud-init to speak with the fabric through waagent. By removing this patch, it will now use the builtin functionality for fabric negotiation.

### CODE REVIEW

**Comment by trstringer** [comment]
@raharper I've actually already validated this with [this script](https://gist.github.com/trstringer/15629fe7ceea7d48b20705cd5fb4d0b4). Let me know if you need more validation though!

**Comment by raharper** [comment]
Thanks! @trstringer   A couple more scenarios worth testing:

1) reboot after initial boot, cloud-init takes a slightly different path on reboot than first boot
2) upgrade existing xenial instance to new cloud-init

**Comment by trstringer** [comment]
Thanks, @raharper! For number 1, I have modified [my validation script](https://gist.github.com/trstringer/15629fe7ceea7d48b20705cd5fb4d0b4#file-validate_xenial_builtin_agent-sh-L33-L42) to include this reboot check.

But for number 2, what exactly do you mean? Upgrade cloud-init on Xenial? From what version to what version? This build I have right now is a one-off deb that I packaged up for testing. Not sure I'm understanding what you mean.

**Comment by raharper** [comment]
> Thanks, @raharper! For number 1, I have modified [my validation script](https://gist.github.com/trstringer/15629fe7ceea7d48b20705cd5fb4d0b4#file-validate_xenial_builtin_agent-sh-L33-L42) to include this reboot check.
> 
> But for number 2, what exactly do you mean? Upgrade cloud-init on Xenial? From what version to what version? This build I have right now is a one-off deb that I packaged up for testing. Not sure I'm understanding what you mean.

Launch an existing Ubuntu 16.04 image (release and daily) and in each one, ssh into it after it's booted, copy in the deb you've made, and dpkg --install it ; this is equivalent to existing instances on a cloud getting an cloud-init update from the archive.  We want to verify that running instances in the cloud today will work after upgrading to this change (versus a newly booted instance with the change).

**Comment by trstringer** [comment]
Let me know what you think about these additional validations!

* [Additional validation](https://gist.github.com/trstringer/15629fe7ceea7d48b20705cd5fb4d0b4#file-validate_xenial_builtin_agent-sh-L16-L24) to reboot and verify after installation on existing VM.
* [Added validation for daily image](https://gist.github.com/trstringer/15629fe7ceea7d48b20705cd5fb4d0b4#file-validate_xenial_builtin_agent-sh-L55)

**Comment by raharper** [comment]
Generally looks, wondering if any of those paths does a cloud-init --clean to have cloud-init take some of the first boot paths again, like a new instance would.

**Comment by raharper** [comment]
We also do some additional sanity checks when we do our scripted sru verifications:

https://github.com/cloud-init/ubuntu-sru/blob/master/sru-templates/manual/azure-sru

**Comment by trstringer** [comment]
> Generally looks, wondering if any of those paths does a cloud-init --clean to have cloud-init take some of the first boot paths again, like a new instance would.

Nope they don't. But I've [added that to my validation script](https://gist.github.com/trstringer/15629fe7ceea7d48b20705cd5fb4d0b4#file-validate_xenial_builtin_agent-sh-L43), so that scenario should be covered now!

**Comment by trstringer** [comment]
> We also do some additional sanity checks when we do our scripted sru verifications:
> 
> https://github.com/cloud-init/ubuntu-sru/blob/master/sru-templates/manual/azure-sru

Ah good to know. I'll run those as well! I'll let you know how that turns out.

**Comment by trstringer** [comment]
What exactly are we looking for particularly in [this grep](https://github.com/cloud-init/ubuntu-sru/blob/master/sru-templates/manual/azure-sru#L109)? My validation is failing on this. What log message should be present that includes "Trace"?

**Comment by raharper** [comment]
> What exactly are we looking for particularly in [this grep](https://github.com/cloud-init/ubuntu-sru/blob/master/sru-templates/manual/azure-sru#L109)? My validation is failing on this. What log message should be present that includes "Trace"?

You can accept non-zero value;  we're looking to see if python had a Traceback error;  not finding it is expected and OK.

**Comment by trstringer** [comment]
Ah ok good to know. Is that the only command that we expect (and want) a non-zero return code? Looking at the rest, I think that's a "yes", but just wanted to confirm.

**Comment by raharper** [comment]
> Ah ok good to know. Is that the only command that we expect (and want) a non-zero return code? Looking at the rest, I think that's a "yes", but just wanted to confirm.

Yes, the rest should all return zero.

**Comment by trstringer** [comment]
@raharper sorry for the delay! I have translated your SRU validation to [do the same validations](https://gist.github.com/trstringer/e06848afa6072b3fd8617856440c6599). And the [output can be found here](https://gist.github.com/trstringer/b4094c4274d55515c72c8bfc621c7a50). Let me know if you think of anything else before we move forward with removing this patch!

**Comment by raharper** [comment]
Thanks for the update.  ISTR some issues (discussed in IRC) with the change; were those transient errors or was there something more serious?  Could you clarify what you ran into and why it's not an issue?

**Comment by trstringer** [comment]
Yep! What the problem was that I had built the cloud-init deb in my development environment (Debian Testing). And because of that, when I installed/upgraded cloud-init on Ubuntu and cleaned/rebooted, it picked up the distro as Debian instead of Ubuntu.

*That* was a problem because the network config for Debian has no file extension and the one for Ubuntu has a .cfg file extension. That caused issues because the interfaces config pulled in only *.cfg files for Ubuntu, so the Debian-like network config was not applied.

In summary, it was how I built cloud-init. By pushing the source pkg to my ppa I was able to get around that.

**Comment by trstringer** [comment]
@raharper thanks again for the review! Did you want to get more people from cloud-init or more from Azure to look at this?

**Comment by OddBloke** [comment]
Hey @trstringer, thanks for the follow-up! Ryan's out for the rest of the year, and we'll all be following him at the end of this week. We have a couple of internal stakeholders that we need to check this with (and they will probably want to run it past their Azure counterparts), so we'll pick this up in the new year!

**Comment by trstringer** [comment]
Sounds good, @OddBloke!!

**Comment by trstringer** [comment]
@OddBloke @raharper I just wanted to check in on this PR. Let me know if you are waiting for anything from my side!

**Comment by raharper** [comment]
It looks like you need to rebase this branch since we started the last SRU.  After a rebase  I believe we're ready to land this.

**Comment by trstringer** [comment]
@raharper done!
